loopout.go

```diff
@@ -638,50 +664,116 @@ func (s *loopOutSwap) payInvoiceAsync(ctx context.Context,
 	paymentStateCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	payStatusChan, payErrChan, err := s.lnd.Router.SendPayment(
-		paymentStateCtx, req,
+	// Attempt to acquire and initialize the routing plugin.
+	routingPlugin, err := AcquireRoutingPlugin(
+		ctx, pluginType, *s.lnd, target, amt,
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	for {
-		select {
-		// Payment advanced to the next state.
-		case payState := <-payStatusChan:
-			s.log.Infof("Payment %v: %v", hash, payState)
-
-			switch payState.State {
-			case lnrpc.Payment_SUCCEEDED:
-				return &payState, nil
-
-			case lnrpc.Payment_FAILED:
-				return &payState, nil
+	var (
+		start          time.Time
+		tryCount       int
+		paymentSuccess bool
+	)
+	if routingPlugin != nil {
+		s.log.Infof("Acquired routing plugin %v for payment %v",
+			pluginType, hash.String())
+
+		defer func() {
+			ReleaseRoutingPlugin(ctx)
+			dt := time.Since(start)
+			if err := s.swapKit.server.ReportRoutingResult(
+				ctx, s.swapInfo().SwapHash, pluginType,
+				paymentSuccess, int32(tryCount),
+				dt.Milliseconds(),
+			); err != nil {
+				s.log.Warnf("Failed to report routing result: "+
+					"%v", err)
+			}
+		}()
+	}
 
-			case lnrpc.Payment_IN_FLIGHT:
-				// Continue waiting for final state.
+	// canRetry returns true if the payment timed out and the
+	// we haven't attempted the payment too many times.
+	canRetry := func(tryCount int, status lndclient.PaymentStatus) bool {
+		return (tryCount <= maxRetries &&
+			status.FailureReason ==
+				lnrpc.PaymentFailureReason_FAILURE_REASON_TIMEOUT)
+	}
 
-			default:
-				return nil, errors.New("unknown payment state")
-			}
+	start = time.Now()
```

Comment:
Updating to make it slower and doing 3x 20 sounds ok to me.

There are a lot of variables so I think the benefit of extending pathfinding time is to make it be at least no worse than the existing pathfinding at finding a route since the new heuristics would only start working as a last resort.